lib/cisco_node_utils/ace_v4.rb

```diff
@@ -0,0 +1,147 @@
+    attr_reader :acl_name, :seqno, :action, :proto, :v4_src_addr_format, :v4_dst_addr_format
+    attr_reader :v4_src_port_format, :v4_dst_port_format, :option_format
+
+    # name: name of the router instance
+    # instantiate: true = create router instance
+    def initialize(acl_name, seqno, action, proto, v4_src_addr_format,
+                   v4_dst_addr_format, v4_src_port_format, v4_dst_port_format,
+                   option_format='', instantiate=true)
+      addr_regex = /(any|host \d+\.\d+\.\d+\.\d+|\d+\.\d+\.\d+\.\d+\/\d+|\d+\.\d+\.\d+\.\d+ \d+\.\d+\.\d+\.\d+)/
+      addr_port_regex = /((eq|neq|lt|gt)\s+\d+|range\s+\d+\s+\d+){0,1}/
+      action_regex = /(permit|deny)/
+      proto_regex = /(sctp|tcp|ip|udp|pim|icmp|igmp|gre|nos|pcp|ospf|eigrp|ahp|esp|ipv6)/
+      fail ArgumentError.new("acl name len must be greater than 0") unless acl_name.length > 0
+      fail ArgumentError.new("seqno must be an integer") unless seqno.is_a?(Integer)
+      fail ArgumentError.new("action has to be permit or deny") unless action_regex.match(action)
+      fail ArgumentError.new("proto is invalid") unless proto_regex.match(proto)
+      fail ArgumentError.new("invalid src addr #{v4_src_addr_format}") unless addr_regex.match(v4_src_addr_format)
+      fail ArgumentError.new("invalid dst addr") unless addr_regex.match(v4_dst_addr_format)
+      fail ArgumentError.new("invalid src port") unless addr_port_regex.match(v4_src_port_format)
+      fail ArgumentError.new("invalid dst port") unless addr_port_regex.match(v4_dst_port_format)
+
+      @acl_name = acl_name
+      @seqno    = seqno
+      @action   = action
+      @proto    = proto
+      @v4_src_addr_format = v4_src_addr_format
+      @v4_dst_addr_format = v4_dst_addr_format
+      @v4_src_port_format = v4_src_port_format
+      @v4_dst_port_format = v4_dst_port_format
+      @option_format      = option_format
+      @set_args = {} 
+      @get_args = {} 
+      create if instantiate
+    end
+
+    # Create a hash of all current router instances.
+    def self.l3_aces
+      instances = config_get('ace_v4' , 'l3_ace')
+      return {} if instances.nil?
+      hash = {}
+      instances.each do |name|
+        hash[name] = RouterAce.new(acl_name, false)
+      end
+      return hash
+    rescue Cisco::CliError => e
+      # CLI will syntax reject when feature is not enabled
+      raise unless e.clierror =~ /Syntax error/
+      return {}
+    end
+    
+    # Create a hash of all current router instances.
+    def self.l4_aces
+      instances = config_get('ace_v4' , 'l4_ace')
+      return {} if instances.nil?
+      hash = {}
+      instances.each do |name|
+        hash[name] = RouterAce.new(acl_name, false)
+      end
+      return hash
+    rescue Cisco::CliError => e
+      # CLI will syntax reject when feature is not enabled
+      raise unless e.clierror =~ /Syntax error/
+      return {}
+    end
+
+    def is_l4_ace?
+        "tcp|udp|sctp".match(@proto) ? true : false
```

Comment:
I left a general note about rubocop, but it will flag things like indention.  Here you indent more then 2 spaces.